sympy/core/numbers.py

```diff
@@ -1021,6 +1021,8 @@ class Float(Number):
 
     is_Float = True
 
+    _remove_non_digits = str.maketrans(dict.fromkeys("-+_."))
```

Comment:
Should this include `eE` for exponential notation, too? or maybe better,
```
if num.startswith('_') or not num.translate(cls._remove_non_digits)
```
could be
```
if num.startswith('_') or not all(_.translate(cls._remove_non_digits) for _ in num.lower().split('e'))
```

Reply:
It could be. I was just checking for things that the new `from_str` in mpmath accepts that were previously not accepted:
```python
In [7]: from mpmath.libmp import from_str

In [8]: from_str("_1", 10)
Out[8]: (0, 1, 0, 1)

In [9]: from_str("e1", 10)
---------------------------------------------------------------------------
ValueError
```
Feel free to push a more comprehensive test.

Reply:
Which version allows In[8] to succeed? With 1.3.0 it fails (as I hope it should, since underscore is used to separate digits and if it is present without a bounding digit on each side it could be an entry error):
```python
>>> from mpmath.libmp import from_str
>>> from_str('_1',10)
Traceback (most recent call last):
...
ValueError: could not convert string to float: '_1'
```

Reply:
@oscarbenjamin , the question above was my only remaining question before committing this. An underscore prefaced number should not be translated as a number since this is a valid python identifier. SymPy (with this PR) will no loger allow a terminal underscore on a number since this is ambiguous: underscores should come between digits.